Add `SnifferSV1` to ITF

jbesraa opened the pull request:
resolves #1564

plebhash commented:
oops, looks like adding c8a721bd48f918ce684fda7e0320245ae67f8a98 to #1551 introduced a conflict here

perhaps it would have been better to have merged this first, sorry about that

plebhash commented on roles/roles-utils/network-helpers/src/sv1_connection.rs:
```diff
@@ -0,0 +1,171 @@
+
+/// Represents a connection between two roles communicating using SV1 protocol.
+///
+/// This struct can be used read and write messages to the other side of the connection.  The
+/// channel is unidirectional, so you will need two instances of this struct to communicate in both
+/// directions.
```
this comment feels a bit confusing

with `sender` I can send, with `receiver` I can receive... therefore this is bidirectional?

frankomosh replied:
from my understanding I think they say a single Channel instance only handles communication in one direction. But yes I agree that the statement could be clearer.

jbesraa replied:
@frankomosh is right. 
Updated the docs. Let me know if it is still not clear.

plebhash commented on roles/roles-utils/network-helpers/Cargo.toml:
```diff
@@ -20,12 +20,16 @@ tokio = { version = "1", features = ["full"] }
 [package.metadata.docs.rs]
-features = ["with_buffer_pool"]
\ No newline at end of file
+features = ["with_buffer_pool"]
```
we should add `sv1` feature here so it is added to docs

plebhash commented on roles/test-utils/mining-device-sv1/src/client.rs:
```diff
@@ -240,14 +269,21 @@ impl Client {
-        }
+        tokio::select!(
+            _ = tokio::signal::ctrl_c() => {
+                warn!("Stopping sv1 miner");
```
we just log a warning on ctrl+c?

plebhash commented on roles/test-utils/mining-device-sv1/src/client.rs:
```diff
@@ -196,6 +222,9 @@ impl Client {
                 warn!("Stopping miner")
               },
+              _ = tokio::signal::ctrl_c() => {
+                  info!("Stopping miner");
```
we just print a info log on ctrl+c?

plebhash commented on roles/test-utils/mining-device-sv1/src/client.rs:
```diff
@@ -106,34 +122,44 @@ impl Client {
-                            error!("Failed to send message to receiver_incoming: {:?}", e);
-                            break; // Exit the loop if sending fails
+            tokio::select!(
+                _ = tokio::signal::ctrl_c() => { },
```
is this intentionally empty?

plebhash replied:
sorry dumb question.. just refreshed my understanding of `tokio::select` and now I understand that as soon as ctrl+c happens, all the other branches are cancelled

resolving the similar questions I asked in other parts of the PR

plebhash commented on test/integration-tests/lib/sv1_sniffer.rs:
```diff
@@ -0,0 +1,203 @@
+    pub async fn wait_for_message_from_downstream(&self, message: &str) {
+        let now = std::time::Instant::now();
+        tokio::select!(
+            _ = tokio::signal::ctrl_c() => { },
```
is this intentionally empty?

plebhash commented on test/integration-tests/lib/sv1_sniffer.rs:
```diff
@@ -0,0 +1,203 @@
+    pub async fn wait_for_message_from_upstream(&self, message: &[&str]) {
+        let now = std::time::Instant::now();
+        tokio::select!(
+            _ = tokio::signal::ctrl_c() => { },
```
is this intentionally empty?

plebhash commented on test/integration-tests/lib/sv1_sniffer.rs:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /// Wait for a specific message to be received from the downstream role.
+    pub async fn wait_for_message_from_downstream(&self, message: &str) {
```
I see this is diverging from the Sv2 Sniffer `wait_for_message_type` pattern

here we have two distinct functions, while on Sv2 Sniffer we just have one and take `MessageDirection` as input

is there some fundamental limitation that forces this specific design?

if not, I think maybe it's better to have both following the same patterns, just to make the APIs a bit more intutive

plebhash replied:
as discussed in call with @jbesraa , there are some fundamental limitations coming from `sv1_api::json_rpc` module that are imposing this pattern

plebhash commented on test/integration-tests/lib/sv1_sniffer.rs:
```diff
@@ -0,0 +1,203 @@
+                .await
+                .map_err(|_| SnifferError::DownstreamClosed)?;
+            upstream_messages.add_message(msg.clone()).await;
+            sleep(Duration::from_millis(100)).await;
```
why sleep here?

plebhash commented on test/integration-tests/Cargo.toml:
```diff
@@ -31,12 +31,17 @@ key-utils = { path = "../../utils/key-utils" }
 path = "lib/mod.rs"
+
+[features]
+default = []
```
```suggestion
default = ["sv1"]
```

having to add `--features sv1` to `cargo test` feels a bit counter-intuitive... and forgetting to do it basically disables the tests that use Sv1 Sniffer

plebhash replied:
as discussed in a call with @jbesraa , we still only have one single test where we assert any kind of Sv1 behavior

and including the `sv1` feature flag as default would bloat the ITF build process for every other test

so for now, it's ok to leave this as not being a default feature

plebhash commented on test/integration-tests/lib/sv1_sniffer.rs:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /// Wait for a specific message to be received from the upstream role.
+    pub async fn wait_for_message_from_upstream(&self, message: &[&str]) {
```
also this takes an array, while `wait_for_message_from_downstream` doesn't

which is yet another reason I believe these should be just one function instead of two (to avoid fragmented patterns)

plebhash replied:
as discussed in call with @jbesraa , there are some fundamental limitations coming from sv1_api::json_rpc module that are imposing this pattern

plebhash commented:
I added some tracing logs, please cherry-pick in case you agree

ea02800241e217b137670783c18ec4bc20342e4d

58feb8a66ee419795d5bf06a69f91b3f542647ba

jbesraa commented:
> I added some tracing logs, please cherry-pick in case you agree
> 
> [ea02800](https://github.com/stratum-mining/stratum/commit/ea02800241e217b137670783c18ec4bc20342e4d)
> 
> [58feb8a](https://github.com/stratum-mining/stratum/commit/58feb8a66ee419795d5bf06a69f91b3f542647ba)

Squashed the SV1 related one to SV1 commit and  picked SV2 commit in

plebhash commented on test/integration-tests/tests/sv1.rs:
```diff
@@ -0,0 +1,27 @@
+            ],
+            MessageDirection::ToDownstream,
+        )
+        .await;
```
looks like we're missing `subscribe`, `authorize` and `notify`

`notify` is not on the original MG test, but it's still worth adding since it will happen on this setup
